forge/lib/oct_encoding.py

```diff
@@ -0,0 +1,53 @@
+
+# -*- coding: utf-8 -*-
+
+
+import forge.lib.cartesian3d as c3d
+
+
+def clamp(val, minVal, maxVal):
+    return max(min(val, maxVal), minVal)
+
+
+def signNotZero(v):
+    return -1.0 if v > 0.0 else 1.0
+
+
+# Converts a scalar value in the range [-1.0, 1.0] to a 8-bit 2's complement number.
+def toSnorm(v):
+    return round((clamp(v, -1.0, 1.0) * 0.5 + 0.5) * 255.0)
+
+
+def fromSnorm(v):
+    return clamp(v, 0.0, 255.0) / 255.0 * 2.0 - 1.0
+
+
+# Compress x, y, z 96-bit floating point into x, z 16-bit representation (2 snorm values)
+def octEncode(vec):
+    res = [0.0, 0.0]
+    l1Norm = float(abs(vec[0]) + abs(vec[1]) + abs(vec[2]))
+    res[0] = vec[0] / l1Norm
+    res[1] = vec[1] / l1Norm
+
+    if vec[2] < 0.0:
+        x = res[0]
+        y = res[1]
+        res[0] = (1.0 - abs(y)) * signNotZero(x)
+        res[1] = (1.0 - abs(x)) * signNotZero(y)
+
+    res[0] = int(toSnorm(res[0]))
+    res[1] = int(toSnorm(res[1]))
+    return res
+
+
+def octDecode(x, y):
+    res = [x, y, 0.0]
+    res[0] = fromSnorm(x)
+    res[1] = fromSnorm(y)
+    res[2] = 1.0 - (abs(res[1]) - abs(res[1]))
+
+    if res[2] < 0.0:
+        oldX = res[0]
+        res[0] = (1.0 - abs(res[1]) * signNotZero(oldX))
+        res[1] = (1.0 - abs(oldX) * signNotZero(res[1]))
+    return c3d.normalize(res)
```

Comment:
Is this encoding/decoding based on https://github.com/AnalyticalGraphicsInc/cesium/blob/b161b6429b9201c99e5fb6f6e6283f3e8328b323/Source/Core/AttributeCompression.js#L43 ? Or did you use a different source?